File: backend/models/messages.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// ConversationType represents the type of a conversation.
type ConversationType string

const (
	ConversationTypeGlobal        ConversationType = "global"
	ConversationTypeTeam          ConversationType = "team"
	ConversationTypeGroup         ConversationType = "group"
	ConversationTypeDM            ConversationType = "dm"
	ConversationTypeSelf          ConversationType = "self"
	ConversationTypePasswordVault ConversationType = "password_vault"
)

// ConversationMemberRole represents the role of a user in a conversation.
type ConversationMemberRole string

const (
	ConversationMemberRoleOwner  ConversationMemberRole = "owner"
	ConversationMemberRoleAdmin  ConversationMemberRole = "admin"
	ConversationMemberRoleMember ConversationMemberRole = "member"
	ConversationMemberRoleViewer ConversationMemberRole = "viewer"
)

// SuggestionStatus is the lifecycle state of a message suggestion.
type SuggestionStatus string

const (
	SuggestionStatusPending   SuggestionStatus = "pending"
	SuggestionStatusAccepted  SuggestionStatus = "accepted"
	SuggestionStatusDismissed SuggestionStatus = "dismissed"
)

// Conversation is a user-to-user chat space (global/team/group/dm/self/password).
type Conversation struct {
	ID        uint           `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`

	Type       ConversationType `json:"type" gorm:"not null;index"`
	Name       string           `json:"name" gorm:"not null"`
	Topic      string           `json:"topic"`
	TeamID     *uint            `json:"team_id,omitempty" gorm:"index"`
	Team       *Team            `json:"team,omitempty" gorm:"foreignKey:TeamID"`
	CreatedBy  uint             `json:"created_by" gorm:"not null;index"`
	Creator    User             `json:"creator,omitempty" gorm:"foreignKey:CreatedBy"`
	IsDefault  bool             `json:"is_default" gorm:"default:false;index"`
	IsArchived bool             `json:"is_archived" gorm:"default:false;index"`

	LastMessageAt *time.Time `json:"last_message_at"`

	Members  []ConversationMember `json:"members,omitempty" gorm:"foreignKey:ConversationID"`
	Messages []Message            `json:"messages,omitempty" gorm:"foreignKey:ConversationID"`
}

// ConversationMember links users to conversations.
type ConversationMember struct {
	ID        uint           `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`

	ConversationID uint                   `json:"conversation_id" gorm:"not null;index:idx_conv_member,unique"`
	UserID         uint                   `json:"user_id" gorm:"not null;index:idx_conv_member,unique"`
	Role           ConversationMemberRole `json:"role" gorm:"not null;default:member"`
	JoinedAt       time.Time              `json:"joined_at"`

	LastReadMessageID *uint      `json:"last_read_message_id,omitempty" gorm:"index"`
	LastReadAt        *time.Time `json:"last_read_at,omitempty"`
	MutedUntil        *time.Time `json:"muted_until,omitempty"`
	IsHidden          bool       `json:"is_hidden" gorm:"default:false"`

	Conversation Conversation `json:"conversation,omitempty" gorm:"foreignKey:ConversationID"`
	User         User         `json:"user,omitempty" gorm:"foreignKey:UserID"`
}

// Message is a single chat message in a conversation.
type Message struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	ConversationID uint         `json:"conversation_id" gorm:"not null;index"`
	Conversation   Conversation `json:"conversation,omitempty" gorm:"foreignKey:ConversationID"`
	SenderID       uint         `json:"sender_id" gorm:"not null;index"`
	Sender         User         `json:"sender,omitempty" gorm:"foreignKey:SenderID"`

	Body        string     `json:"body" gorm:"type:text"`
	IsSensitive bool       `json:"is_sensitive" gorm:"default:false"`
	EditedAt    *time.Time `json:"edited_at,omitempty"`
	DeletedAt   *time.Time `json:"deleted_at,omitempty" gorm:"index"`

	MetadataJSON string `json:"metadata_json" gorm:"type:text"`

	Attachments []MessageAttachment `json:"attachments,omitempty" gorm:"foreignKey:MessageID"`
	References  []MessageReference  `json:"references,omitempty" gorm:"foreignKey:MessageID"`
	Suggestions []MessageSuggestion `json:"suggestions,omitempty" gorm:"foreignKey:MessageID"`
	Reactions   []MessageReaction   `json:"reactions,omitempty" gorm:"foreignKey:MessageID"`
}

// MessageAttachment represents file/link-style message attachments.
type MessageAttachment struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	MessageID uint    `json:"message_id" gorm:"not null;index"`
	Message   Message `json:"message,omitempty" gorm:"foreignKey:MessageID"`

	Kind        string `json:"kind" gorm:"not null;index"` // file,image,youtube,github,website,bookmark,task,event,calendar,activity,learning_path,saved_search,voice_note
	FileID      *uint  `json:"file_id,omitempty" gorm:"index"`
	URL         string `json:"url"`
	Title       string `json:"title"`
	PreviewJSON string `json:"preview_json" gorm:"type:text"`
}

// MessageReference maps chat messages to Trackeep entities.
type MessageReference struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	MessageID uint    `json:"message_id" gorm:"not null;index"`
	Message   Message `json:"message,omitempty" gorm:"foreignKey:MessageID"`

	EntityType string `json:"entity_type" gorm:"not null;index"`
	EntityID   uint   `json:"entity_id" gorm:"not null;index"`
	DeepLink   string `json:"deep_link" gorm:"not null"`
}

// MessageSuggestion stores non-blocking smart suggestions triggered by message text.
type MessageSuggestion struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	MessageID uint    `json:"message_id" gorm:"not null;index"`
	Message   Message `json:"message,omitempty" gorm:"foreignKey:MessageID"`

	Type        string           `json:"type" gorm:"not null;index"` // create_task, create_event, save_bookmark, ...
	PayloadJSON string           `json:"payload_json" gorm:"type:text"`
	Status      SuggestionStatus `json:"status" gorm:"not null;default:pending;index"`
}

// MessageReaction stores emoji reactions.
type MessageReaction struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	MessageID uint    `json:"message_id" gorm:"not null;index:idx_message_reaction,unique"`
	Message   Message `json:"message,omitempty" gorm:"foreignKey:MessageID"`
	UserID    uint    `json:"user_id" gorm:"not null;index:idx_message_reaction,unique"`
	User      User    `json:"user,omitempty" gorm:"foreignKey:UserID"`
	Emoji     string  `json:"emoji" gorm:"not null;index:idx_message_reaction,unique"`
}

// PasswordVaultItem is encrypted secret data owned by a user.
type PasswordVaultItem struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	OwnerUserID uint `json:"owner_user_id" gorm:"not null;index"`
	OwnerUser   User `json:"owner_user,omitempty" gorm:"foreignKey:OwnerUserID"`

	Label           string     `json:"label" gorm:"not null"`
	EncryptedSecret string     `json:"-" gorm:"type:text;not null"`
	EncryptedNotes  string     `json:"-" gorm:"type:text"`
	SourceMessageID *uint      `json:"source_message_id,omitempty" gorm:"index"`
	SourceMessage   *Message   `json:"source_message,omitempty" gorm:"foreignKey:SourceMessageID"`
	CreatedBy       uint       `json:"created_by" gorm:"not null;index"`
	LastAccessedAt  *time.Time `json:"last_accessed_at,omitempty"`

	Shares []PasswordVaultShare `json:"shares,omitempty" gorm:"foreignKey:VaultItemID"`
}

// PasswordVaultShare controls explicit sharing of vault items.
type PasswordVaultShare struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	VaultItemID uint              `json:"vault_item_id" gorm:"not null;index"`
	VaultItem   PasswordVaultItem `json:"vault_item,omitempty" gorm:"foreignKey:VaultItemID"`

	SharedByUserID       uint         `json:"shared_by_user_id" gorm:"not null;index"`
	SharedByUser         User         `json:"shared_by_user,omitempty" gorm:"foreignKey:SharedByUserID"`
	TargetConversationID uint         `json:"target_conversation_id" gorm:"not null;index"`
	TargetConversation   Conversation `json:"target_conversation,omitempty" gorm:"foreignKey:TargetConversationID"`
	ExpiresAt            *time.Time   `json:"expires_at,omitempty"`
	AllowReveal          bool         `json:"allow_reveal" gorm:"default:false"`
}

File: frontend/src/lib/messages.ts
export type ConversationType = 'global' | 'team' | 'group' | 'dm' | 'self' | 'password_vault';

export interface UserLite {
  id: number;
  username: string;
  full_name?: string;
  avatar_url?: string;
}

export interface Conversation {
  id: number;
  type: ConversationType;
  name: string;
  topic?: string;
  team_id?: number | null;
  created_by: number;
  is_default: boolean;
  is_archived: boolean;
  last_message_at?: string;
  created_at: string;
  updated_at: string;
}

export interface ConversationListItem {
  conversation: Conversation;
  role: string;
  unread_count: number;
  last_message?: Message;
}

export interface ConversationMember {
  id: number;
  conversation_id: number;
  user_id: number;
  role: 'owner' | 'admin' | 'member' | 'viewer' | string;
  joined_at?: string;
  last_read_message_id?: number | null;
  last_read_at?: string | null;
  muted_until?: string | null;
  is_hidden?: boolean;
  user?: UserLite;
}

export interface MessageAttachment {
  id: number;
  message_id: number;
  kind: string;
  file_id?: number | null;
  url?: string;
  title?: string;
  preview_json?: string;
}

export interface MessageReference {
  id: number;
  message_id: number;
  entity_type: string;
  entity_id: number;
  deep_link: string;
}

export interface MessageSuggestion {
  id: number;
  message_id: number;
  type: string;
  payload_json: string;
  status: 'pending' | 'accepted' | 'dismissed';
  created_at: string;
  updated_at: string;
}

export interface MessageReaction {
  id: number;
  message_id: number;
  user_id: number;
  emoji: string;
}

export interface Message {
  id: number;
  conversation_id: number;
  sender_id: number;
  sender?: UserLite;
  body: string;
  is_sensitive: boolean;
  edited_at?: string | null;
  deleted_at?: string | null;
  metadata_json?: string;
  created_at: string;
  updated_at: string;
  attachments?: MessageAttachment[];
  references?: MessageReference[];
  suggestions?: MessageSuggestion[];
  reactions?: MessageReaction[];
}

export interface VaultItem {
  id: number;
  label: string;
  owner_user_id: number;
  source_message_id?: number | null;
  last_accessed_at?: string | null;
  shared: boolean;
  allow_reveal: boolean;
  expires_at?: string | null;
  target_conversation_id?: number | null;
}

export interface UserFile {
  id: number;
  original_name: string;
  mime_type: string;
  file_size: number;
  created_at: string;
  description?: string;
}

export interface WsEvent {
  type: string;
  conversation_id?: number;
  data?: any;
  timestamp?: string;
}

const API_BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:8080';

function getToken() {
  return localStorage.getItem('trackeep_token') || localStorage.getItem('token') || '';
}

async function apiRequest<T>(path: string, options: RequestInit = {}): Promise<T> {
  const token = getToken();
  const res = await fetch(`${API_BASE_URL}/api/v1/messages${path}`, {
    ...options,
    headers: {
      ...(options.body instanceof FormData ? {} : { 'Content-Type': 'application/json' }),
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
      ...(options.headers || {}),
    },
  });
  if (!res.ok) {
    const errorData = await res.json().catch(() => ({}));
    throw new Error(errorData.error || `Request failed (${res.status})`);
  }
  return res.json();
}

export const messagesApi = {
  listConversations: () => apiRequest<{ conversations: ConversationListItem[] }>('/conversations'),
  createConversation: (payload: any) => apiRequest<{ conversation: Conversation }>('/conversations', {
    method: 'POST',
    body: JSON.stringify(payload),
  }),
  getConversation: (id: number) =>
    apiRequest<{ conversation: Conversation; membership: ConversationMember; members: ConversationMember[] }>(
      `/conversations/${id}`
    ),
  getMessages: (conversationId: number, cursor?: number, limit: number = 50) =>
    apiRequest<{ messages: Message[]; next_cursor?: number }>(
      `/conversations/${conversationId}/messages?limit=${limit}${cursor ? `&cursor=${cursor}` : ''}`
    ),
  sendMessage: (conversationId: number, payload: any) => apiRequest<{ message: Message; warning?: string }>(
    `/conversations/${conversationId}/messages`,
    {
      method: 'POST',
      body: JSON.stringify(payload),
    }
  ),
  updateMessage: (id: number, body: string) =>
    apiRequest<{ message: Message }>(`/messages/${id}`, { method: 'PATCH', body: JSON.stringify({ body }) }),
  deleteMessage: (id: number) => apiRequest<{ message: string }>(`/messages/${id}`, { method: 'DELETE' }),
  addReaction: (id: number, emoji: string) =>
    apiRequest<{ reaction: MessageReaction }>(`/messages/${id}/reactions`, {
      method: 'POST',
      body: JSON.stringify({ emoji }),
    }),
  removeReaction: (id: number, emoji: string) =>
    apiRequest<{ message: string }>(`/messages/${id}/reactions/${encodeURIComponent(emoji)}`, { method: 'DELETE' }),
  searchMessages: (payload: any) =>
    apiRequest<{ results: Message[]; total: number; limit: number; offset: number }>(
      '/messages/search',
      {
        method: 'POST',
        body: JSON.stringify(payload),
      }
    ),
  getSuggestions: (messageId: number) => apiRequest<{ suggestions: MessageSuggestion[] }>(`/messages/${messageId}/suggestions`),
  acceptSuggestion: (messageId: number, suggestionId: number, payload: any = {}) =>
    apiRequest<any>(`/messages/${messageId}/suggestions/${suggestionId}/accept`, {
      method: 'POST',
      body: JSON.stringify(payload),
    }),
  dismissSuggestion: (messageId: number, suggestionId: number) =>
    apiRequest<any>(`/messages/${messageId}/suggestions/${suggestionId}/dismiss`, {
      method: 'POST',
      body: JSON.stringify({}),
    }),
  revealSensitiveMessage: (messageId: number) =>
    apiRequest<{ message_id: number; plaintext: string }>(`/messages/${messageId}/reveal-sensitive`, {
      method: 'POST',
      body: JSON.stringify({}),
    }),
  listVaultItems: () => apiRequest<{ items: VaultItem[] }>('/password-vault/items'),
  createVaultItem: (payload: any) => apiRequest<any>('/password-vault/items', {
    method: 'POST',
    body: JSON.stringify(payload),
  }),
  shareVaultItem: (id: number, payload: any) =>
    apiRequest<any>(`/password-vault/items/${id}/share`, {
      method: 'POST',
      body: JSON.stringify(payload),
    }),
  revealVaultItem: (id: number) =>
    apiRequest<{ id: number; label: string; secret: string; notes: string; warning?: string }>(
      `/password-vault/items/${id}/reveal`,
      { method: 'POST', body: JSON.stringify({}) }
    ),
  unshareVaultItem: (id: number, payload: any) =>
    apiRequest<any>(`/password-vault/items/${id}/unshare`, {
      method: 'POST',
      body: JSON.stringify(payload),
    }),
  listUserFiles: async (query: string = '', limit: number = 20): Promise<UserFile[]> => {
    const token = getToken();
    const params = new URLSearchParams();
    if (query.trim()) {
      params.set('q', query.trim());
    }
    if (Number.isFinite(limit) && limit > 0) {
      params.set('limit', String(Math.min(100, Math.floor(limit))));
    }

    const suffix = params.toString() ? `?${params.toString()}` : '';
    const res = await fetch(`${API_BASE_URL}/api/v1/files${suffix}`, {
      headers: {
        ...(token ? { Authorization: `Bearer ${token}` } : {}),
      },
    });
    if (!res.ok) {
      const errorData = await res.json().catch(() => ({}));
      throw new Error(errorData.error || `Request failed (${res.status})`);
    }
    return res.json();
  },
};

export async function uploadChatFile(file: File): Promise<{ id: number; original_name: string; mime_type: string }> {
  const token = getToken();
  const formData = new FormData();
  formData.append('file', file);
  formData.append('description', 'Uploaded from chat');

  const res = await fetch(`${API_BASE_URL}/api/v1/files/upload`, {
    method: 'POST',
    headers: {
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: formData,
  });

  if (!res.ok) {
    const errorData = await res.json().catch(() => ({}));
    throw new Error(errorData.error || `Upload failed (${res.status})`);
  }
  return res.json();
}

export class MessagesRealtimeClient {
  private ws: WebSocket | null = null;
  private reconnectTimer: number | null = null;
  private shouldReconnect = true;
  private onEvent: (event: WsEvent) => void;
  private onStatus?: (status: 'connected' | 'disconnected' | 'error') => void;

  constructor(
    onEvent: (event: WsEvent) => void,
    onStatus?: (status: 'connected' | 'disconnected' | 'error') => void
  ) {
    this.onEvent = onEvent;
    this.onStatus = onStatus;
  }

  connect() {
    this.cleanupReconnect();
    const token = getToken();
    if (!token) return;

    const wsBase = API_BASE_URL.replace(/^http/, 'ws');
    this.ws = new WebSocket(`${wsBase}/api/v1/messages/ws?token=${encodeURIComponent(token)}`);

    this.ws.onopen = () => {
      this.onStatus?.('connected');
    };

    this.ws.onmessage = (evt) => {
      try {
        const parsed = JSON.parse(evt.data) as WsEvent;
        this.onEvent(parsed);
      } catch {
        // ignore invalid payloads
      }
    };

    this.ws.onerror = () => {
      this.onStatus?.('error');
    };

    this.ws.onclose = () => {
      this.onStatus?.('disconnected');
      this.ws = null;
      if (this.shouldReconnect) {
        this.reconnectTimer = window.setTimeout(() => this.connect(), 2000);
      }
    };
  }

  send(payload: any) {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;
    this.ws.send(JSON.stringify(payload));
  }

  disconnect() {
    this.shouldReconnect = false;
    this.cleanupReconnect();
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  private cleanupReconnect() {
    if (this.reconnectTimer) {
      window.clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
  }
}
